Migrate to Swift Package Manager

THISISDINOSAUR opened the pull request:
<!--
Note: This checklist is a reminder of our shared engineering expectations. Feel free to change it, although assigning a GitHub reviewer and the items in bold are required.
-->

Task/Issue URL: https://app.asana.com/0/1187352151074490/1187763920072074
Tech Design URL:
CC: @brindy 

**Description**:
Change dependency management from Carthage to Swift Package Manager. This includes:

- Removing Alamofire since it was no longer required, and using URLSessionDataTasks instead
- Removing Device.swift since it doesn't currently support SPM and we don't really use it
- Necessary updates to remaining dependancies to versions that support SPM. This includes API changes to Lottie and OHHTTPStubs

**Steps to test this PR**:
1. Test building the project works without Carthage, and that all tests pass
2. Test all network calls affected by the switch from Alamo fire, including manually testing the features work, and using Charles Proxy to test that the headers etc are unchanged/as expected. The features include:

* [x] favicons
* [x] autocomplete
* [x] searches 
* [x] pixels
* [x] HTTPS upgrading

<!--
Before submitting a PR, please ensure you have tested the combinations you expect the reviewer to test, then delete configurations you *know* do not need explicit testing.

Using a simulator where a physical device is unavailable is acceptable. 
-->

**Copy Testing**:

* [x] Use of correct apostrophes in new copy, ie `’` rather than `'`

**Orientation Testing**:

* [x] Portrait
* [x] Landscape

**Device Testing**:

* [x] iPhone SE (1st Gen)
* [x] iPhone 8
* [x] iPhone X
* [x] iPad

**OS Testing**:

* [x] iOS 11
* [x] iOS 12
* [x] iOS 13

**Theme Testing**:

* [x] Light theme
* [x] Dark theme

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)

THISISDINOSAUR committed:
Remove device.swift

It's not really needed, and they don't currently support SPM, so it was
blocking migration to SPM

THISISDINOSAUR committed:
Remove alamofire from core classes

It is no longer needed, so alamofire requests have been changed to
URLSessionDataTasks

THISISDINOSAUR committed:
Fix Pixel.fire using incorrect HTTP headers

It was using the default headers, rather than the ones passed to the
function

THISISDINOSAUR committed:
Remove alamofire from tests

THISISDINOSAUR committed:
Add URLComponents asURL() function

To replace the one in alamofire

THISISDINOSAUR committed:
Remove Carthage and add Swift package manager

THISISDINOSAUR committed:
Update lottie usage

THISISDINOSAUR committed:
Update Toast import

THISISDINOSAUR committed:
Update OHHTTPS usage

THISISDINOSAUR committed:
Merge branch 'develop' into feature/elle/migrate-to-swift-package-manager

# Conflicts:
#	DuckDuckGo.xcodeproj/project.pbxproj

THISISDINOSAUR committed:
Remove punnycode from IntegrationTests target

THISISDINOSAUR committed:
Add test to APIRequestTests to verify all default headers

THISISDINOSAUR commented on Core/APIHeaders.swift:
```diff
@@ -34,9 +37,18 @@ public class APIHeaders {
     }
 
     public var defaultHeaders: HTTPHeaders {
```
The default headers should be the same as they were with Alamofire

THISISDINOSAUR commented on Core/APIRequest.swift:
```diff
@@ -34,43 +33,73 @@ public class APIRequest {
                                method: HTTPMethod = .get,
-                               parameters: [String: Any]? = nil,
-                               completion: @escaping APIRequestCompletion) -> Request {
+                               parameters: [String: String]? = nil,
```
I change this to only accept strings since we were only ever using strings, and whilst Alamofire accepted other things, as far as I could tell it wouldn't successfully add any parameter that isn't a string to the URL.

THISISDINOSAUR commented on Core/APIRequest.swift:
```diff
@@ -34,43 +33,73 @@ public class APIRequest {
+            
+            if let error = error {
+                completion(nil, error)
+            } else if let error = httpResponse?.validateStatusCode(statusCode: 200..<300) { 
```
An alternative to this, we could potentially do the status code validation on the appropriate URLSessionDelegate callback, but after discussions with Brindy we decided it's unnecessary.
I made this default to returning the dataTask callback error over the status code error, as I figured it was likely to be more useful than a generic HTTPURLResponseError.invalidStatusCode error, but don't have strong options about it. (I think at any rate it's unlikely that they'll be no error at this point but an invalid status code)

THISISDINOSAUR committed:
Update readme to remove Carthage and add SPM

THISISDINOSAUR committed:
Update test bash scripts to remove Carthage steps

brindy commented on Core/URLExtension.swift:
```diff
@@ -174,3 +174,15 @@ extension URL {
+        case invalidURL(url: String?)
+    }
+
+    public func asURL() throws -> URL {
```
Can we just use the `url` property directly? Then we can just dump this extension.

brindy commented on Core/HTTPSUpgrade.swift:
```diff
@@ -95,17 +94,27 @@ public class HTTPSUpgrade {
-        Alamofire.request(serviceRequest).validate(statusCode: 200..<300).response { response in
-            if let data = response.data {
+
+        let task = session.dataTask(with: serviceRequest) {
```
Would it be possible (and make sense) to re-use the `APIRequest `?

brindy commented on Core/Pixel.swift:
```diff
@@ -272,10 +271,20 @@ public class Pixel {
+        var urlRequest = URLRequest(url: url)
+        urlRequest.allHTTPHeaderFields = headers
+
+        let task = session.dataTask(with: urlRequest) {
```
Again, would it be possible (and make sense) to re-use `APIRequest`?

brindy reviewed (changes requested):
As discussed, great job!  3 comments below, and we need to be careful about handling network responses on the correct thread, which Alamofire seemed to be doing for us.

THISISDINOSAUR committed:
Remove asURL() from URLExtension

THISISDINOSAUR committed:
Add ability to APIRequest to callback on the main thread

THISISDINOSAUR committed:
Add URLResponse to Reponse struct of APIRequest

To allow for checking for cached responses

THISISDINOSAUR committed:
Add method to APIRequest to get URLRequest object for a request

To allow HTTPSUpgrade to lookup a cached response

THISISDINOSAUR committed:
Change Pixel.fire() params from [String: String?] to [String: String]

This is because it was never being given optionals, and gives it the
same interface as APIRequest.request()

THISISDINOSAUR committed:
Change Pixel.fire to use APIRequest.request()

THISISDINOSAUR committed:
Fix APIReqeuest.request callback order

THISISDINOSAUR committed:
Change HTTPSUpdate to use APIRequest.request()

THISISDINOSAUR commented on Core/Pixel.swift:
```diff
@@ -257,7 +256,7 @@ public class Pixel {
     public static func fire(pixel: PixelName,
                             forDeviceType deviceType: UIUserInterfaceIdiom = UIDevice.current.userInterfaceIdiom,
-                            withAdditionalParameters params: [String: String?] = [:],
+                            withAdditionalParameters params: [String: String] = [:],
```
I changed this interface slightly so it matches APIRequest.request() and since we weren't ever using optionals here

THISISDINOSAUR committed:
Remove unnecessary URLComponents extension

brindy reviewed (approved):
Great work @THISISDINOSAUR :) LGTM
